lib/collection/src/shards/local_shard/query.rs

```diff
@@ -75,45 +75,100 @@ impl LocalShard {
-        if request.with_payload.is_required() || request.with_vector.is_enabled() {
+        //
+        // TODO(universal-query): it might make sense to change this approach to fetch payload and vector at the collection level
+        // after the shard results merge, but it requires careful benchmarking
+        let fill_futures = batched_scored_points
+            .into_iter()
+            .zip(all_with_payloads_or_vectors)
+            .map(|(query_response, (with_payload, with_vector))| {
+                self.fill_with_payload_or_vectors(query_response, with_payload, with_vector)
+            });
+
+        let batched_scored_points = futures::future::try_join_all(fill_futures).await?;
+
+        Ok(batched_scored_points)
+    }
+
+    /// Fetches the payload and/or vector if required. This will filter out points if they are deleted between search and retrieve.
+    async fn fill_with_payload_or_vectors(
+        &self,
+        query_response: ShardQueryResponse,
+        with_payload: WithPayloadInterface,
+        with_vector: WithVector,
+    ) -> CollectionResult<ShardQueryResponse> {
+        let filled = if with_payload.is_required() || with_vector.is_enabled() {
             // ids to retrieve (deduplication happens in the searcher)
-            let point_ids = scored_points
+            let point_ids = query_response
                 .iter()
                 .flatten()
                 .map(|scored_point| scored_point.id)
                 .collect::<Vec<PointIdType>>();
 
-            // it might make sense to change this approach to fetch payload and vector at the collection level
-            // after the shard results merge, but it requires careful benchmarking
-            let mut records = SegmentsSearcher::retrieve(
+            let mut records_iter = SegmentsSearcher::retrieve(
                 self.segments(),
                 &point_ids,
-                &WithPayload::from(&request.with_payload),
-                &request.with_vector,
-            )?;
-
-            // update scored points in place
-            for (scored_point, record) in scored_points.iter_mut().flatten().zip(records.iter_mut())
-            {
-                scored_point.payload = record.payload.take();
-                scored_point.vector = record.vector.take();
-            }
-        }
+                &(&with_payload).into(),
+                &with_vector,
+            )?
+            .into_iter()
+            .peekable();
+
+            query_response
+                .into_iter()
+                .map(|intermediate| {
+                    intermediate
+                        .into_iter()
+                        // Points might get deleted between search and retrieve.
+                        // But it's not a problem, because we just filter them out.
+                        .filter_map(|mut scored_point| {
+                            match records_iter.peeking_next(|record| record.id == scored_point.id) {
```

Comment:
This part assumes that the flattened iterator over ids above, has the same order as:
1. The response of `SegmentsSearcher::retrieve` (excluding those points which were not found)
2. Iterating over each intermediate response without flattening.

Reply:
Can we have a debug_assert! for detecting when the order is different?

Reply:
And maybe also if the records iterator contains a point we will not be covering anymore? Or does that not make sense?

If that would happen, this logic would break.